src/hotspot/share/gc/parallel/psCompactionManager.hpp

```diff
@@ -111,7 +111,82 @@ class ParCompactionManager : public CHeapObj<mtGC> {
   static RegionTaskQueueSet* region_task_queues()      { return _region_task_queues; }
   OopTaskQueue*  oop_stack()       { return &_oop_stack; }
 
- public:
+  class MarkingStatsCache : public CHeapObj<mtGC> {
+    constexpr static size_t num_entries = 1024;
+    static_assert(is_power_of_2(num_entries), "inv");
+    static_assert(num_entries > 0, "inv");
+
+    constexpr static size_t entry_mask = num_entries - 1;
+
+    struct CacheEntry {
+      size_t region_id;
+      size_t live_words;
+    };
+
+    CacheEntry entries[num_entries] = {};
+
+    void push(size_t region_id, size_t live_words) {
+      size_t index = (region_id & entry_mask);
+      if (entries[index].region_id == region_id) {
+        // Hit
+        entries[index].live_words += live_words;
+        return;
+      }
+      // Miss
+      if (entries[index].live_words != 0) {
+        evict(index);
+      }
+      entries[index].region_id = region_id;
+      entries[index].live_words = live_words;
+    }
+
+  public:
+    void push(oop obj, size_t live_words) {
+      ParallelCompactData& data = PSParallelCompact::summary_data();
+      const size_t region_size = ParallelCompactData::RegionSize;
+
+      HeapWord* addr = cast_from_oop<HeapWord*>(obj);
+      const size_t start_region_id = data.addr_to_region_idx(addr);
+      const size_t end_region_id = data.addr_to_region_idx(addr + live_words - 1);
+      if (start_region_id == end_region_id) {
+        // Completely inside this region
+        push(start_region_id, live_words);
+        return;
+      }
+
+      // First region
+      push(start_region_id, region_size - data.region_offset(addr));
+
+      // Middle regions; bypass cache
+      for (size_t i = start_region_id + 1; i < end_region_id; ++i) {
+        data.region(i)->set_partial_obj_size(region_size);
+        data.region(i)->set_partial_obj_addr(addr);
+      }
+
+      // Last region; bypass cache
+      const size_t end_offset = data.region_offset(addr + live_words - 1);
+      data.region(end_region_id)->set_partial_obj_size(end_offset + 1);
+      data.region(end_region_id)->set_partial_obj_addr(addr);
```

Comment:
I assume most of the benefit comes from having the cache when the whole object is in a single region. To avoid duplicating the logic around what to do when an object spans multiple regions we could just bail and call `add_object(...)` here: 
```suggestion
      data.add_object(obj, live_words);
```

What do you think about that? Or will this remove a big part of the improvement?